### DISCUSSION TITLE
Rust Support

### DISCUSSION THREAD

**Discussion by JannesBrands** [Open-Agriculture/AgIsoStack-plus-plus#215] [Ideas]
Hey there!
First of all, a big thanks for this project. ISOBUS for everyone was and is unfortunately still a problem. I have worked with the software library of OSB-Ag and CCI myself and I am glad to see such a project, which offers a reasonable option to write an ISOBUS application.

I am currently working with Rust and would like to implement the majority of my ideas and projects with Rust in the future. The advantages are well known and I think it would be cool if this library could find a way to be one step ahead of other proprietary ISOBUS libraries. 

I know that a complete rewrite of this project in Rust would be a lot of work, but maybe we could keep this as an idea/thought for now, since this project is not too far advanced yet and it would be more conceivable at this point than if this project fulfills all the functionalities of the 11783 standard.

Translated with www.DeepL.com/Translator (free version)

**Comment by ad3154** [comment]
Hey there!

I have been wanting to learn Rust, and you're not the only person to ask, so I'm at least a little interested in developing a Rust version in another repository. Most of the applications I need this library for are written in C++ today though, so I'd like to get this C++ version at least to some kind of 1.0 stable state before really considering changing gears or splitting my attention with a Rust version. For what it's worth, I don't think that's particularly far away, as my plan has always been to finish the task controller client and the file server client, do some tidying up, then call that 1.0. Then we can switch from major features more towards full ISO/AEF conformance and bugfixes in subsequent updates, which will allow me some more freedom in how I spend my evenings, haha.

So long story short, yes, I'd like to make a Rust version too when I can if life allows. I think more people might contribute as well if it were Rust based, which would be nice. But my priority at least is the C++ version until it's stable and has those last 2 major features working well.

**Comment by GwnDaan** [comment]
Hi!

I thought I'd also join in here. I've been seeing more and more Rust coming by lately, and as far as I've seen it looks great. Especially the on compile-time memory safety attracts me. Like @ad3154  says, we'd probably be best of releasing a stable version first before switching gears. But I'm definitely down for working on a Rust version as well. I saw that [@Thom-De-Jong](https://github.com/Thom-de-Jong) had started working on a Rust version a while ago, but it seems he already pulled it off GitHub. 

I'm pleasantly surprised how many people are interested in ISOBUS, and it seems to me that perhaps the most powerful thing to do is to bring it all under a common GitHub organization. I know I already touched on this in #175 but it seemed useful to bring it up here as well.  Maybe then we'd have to find another name instead of isobus-plus-plus as well, I don't know.

**Comment by ad3154** [comment]
Yeah I agree with everything @GwnDaan says. Maybe we do need a more general organization name, which can house the Isobus++ branded stuff, plus whatever else we come up with, and is more open to just people in general who are interested in ISOBUS. I also don't really want my personal account to be the home for all this stuff haha. I am open to ideas on naming for sure, I am not very good with names.

**Comment by ad3154** [comment]
Converting this issue to a discussion. We still should come up with an organization name so we can place something like a Rust implementation under that grouping.

**Comment by Thom-de-Jong** [comment]
I have indeed been working on and off on a rust implementation. Unfortunately, due to my personal life, I do not have a lot of time to consistently work on the project.
My old pure Rust implementation, I started 2 years ago, can be found under the name [OpenIsobus](https://github.com/OpenIsobus/OpenIsobus)
The examples should still work but I have not tested them in a while.

I quit work on that pure Rust version in favor of a wrapper around this library. The reasoning behind that decision is that it would be less work to get going.
A con however is that we no longer have the memory guarantees of Rust, which is the main reason to use Rust in the first place!

It might not be to difficult to rewrite the library in Rust, once the API is stable, and keep it in a separate repository, or maybe just make sure the C++ implementation does not have memory problems. 😝
What do you think? @ad3154, @GwnDaan, @JannesBrands

For now you can follow my slow progress in my repo [AgIsoStack-rs](https://github.com/Thom-de-Jong/AgIsoStack-rs)

**Reply by JannesBrands** [reply]
I know in reality that would be a lot of work, but would it not theoretically make more sense to write all the library in pure rust to have a memory safe library and in consequence using interoperability to c++? 
So far I don't know how well it's working accessing rust code from c++, just that in general there is something working.
What do you think of that guys? Does it make sense at least in theory? It's another topic if it does make sense at this moment of progress of the c++ lib.

**Reply by ad3154** [reply]
I think the ultimate goal should be to write it fully in Rust rather than do it as a wrapper once things are "stable", and I think that is what @Thom-de-Jong is saying as well.

**Reply by Thom-de-Jong** [reply]
That is exactly what I meant to say.
If you, @ad3154, agree with taking that direction I would love to start working on the “official” AgIsoStack-rs inside the Open-Agriculture organisation.
As I will need a working Rust wrapper/ implementation at the company I work for in the upcoming year. I will be able to spend some more time on the project as I used to in the last year.

**Reply by JannesBrands** [reply]
Yeah I agree. Lately I am interested to work on isobus in Rust. Would be good to centralize it in this organization. 
Maybe its possible for @ad3154  to set it up in the next days. Depending on the exams those weeks, I will start to contribute.

**Reply by GwnDaan** [reply]
Hi all, I created the project here https://github.com/Open-Agriculture/AgIsoStack-rs for us to get started 😄 

I think the plan initially was to try to convert this C++ project into Rust where possible 1:1. But since we are still in the middle of refactoring the core away from the statics/singletons it might not be possible to convert all of it yet. Nevertheless I think we can make a good start with the basics, e.g. the project layout, control functions, address claiming, hardware abstraction, socketcan implementation, github actions, etc. 

I'll try to create some issues for people to work on throughout this week, and later possibly a roadmap/"project" to track our progress. Feel free to raise any issues yourself if anything is missing and should be added!

Also this discussion about the Rust API design and programming practices we had a month ago, might be a good read for anyone interested: https://github.com/Open-Agriculture/AgIsoStack-plus-plus/discussions/279

**Comment by ad3154** [comment]
Please check out this announcement over in the other repo and see what you all think: https://github.com/Open-Agriculture/AgIsoStack-rs/discussions/1
